pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/timeout/TransactionTimeoutTrackerImpl.java

```diff
@@ -0,0 +1,120 @@
+
+    private final Timer timer;
+    private final TripleLongPriorityQueue priorityQueue = new TripleLongPriorityQueue();
+    private final long tickTimeMillis;
+    private final Clock clock;
+    private final static long BASE_OF_MILLIS_TO_SECOND = 1000L;
+    private Timeout currentTimeout;
+    private final static long INITIAL_TIMEOUT = 1L;
+    private long nowTaskTimeoutTime = INITIAL_TIMEOUT;
+    private final long tcId;
+    private final TransactionMetadataStoreService transactionMetadataStoreService;
+
+    TransactionTimeoutTrackerImpl(long tcId, Timer timer, long tickTimeMillis,
+                                  TransactionMetadataStoreService transactionMetadataStoreService) {
+        this.tcId = tcId;
+        this.transactionMetadataStoreService = transactionMetadataStoreService;
+        this.timer = timer;
+        this.tickTimeMillis  = tickTimeMillis;
+        this.clock = Clock.systemUTC();
+    }
+
+    @Override
+    public CompletableFuture<Boolean> addTransaction(long sequenceId, long timeout) {
+        if (timeout < tickTimeMillis) {
+            this.transactionMetadataStoreService.endTransaction(new TxnID(priorityQueue.peekN2(),
+                    priorityQueue.peekN3()), TxnAction.ABORT_VALUE);
+            return CompletableFuture.completedFuture(false);
+        }
+        synchronized (this){
+            long nowTime = clock.millis() / BASE_OF_MILLIS_TO_SECOND;
+            priorityQueue.add(timeout + nowTime, tcId, sequenceId);
+            long nowTransactionTimeoutTime = nowTime + timeout;
+            if (nowTaskTimeoutTime == INITIAL_TIMEOUT) {
+                currentTimeout = timer.newTimeout(this, timeout, TimeUnit.SECONDS);
+                nowTaskTimeoutTime = nowTransactionTimeoutTime;
+            } else if (nowTaskTimeoutTime > nowTransactionTimeoutTime) {
+                currentTimeout.cancel();
+                currentTimeout = timer.newTimeout(this, timeout, TimeUnit.SECONDS);
+                nowTaskTimeoutTime = nowTransactionTimeoutTime;
+            }
+        }
+        return CompletableFuture.completedFuture(false);
+    }
+
+    @Override
+    public void replayAddTransaction(long sequenceId, long timeout) {
+        long nowTime = clock.millis() / BASE_OF_MILLIS_TO_SECOND;
+        priorityQueue.add(timeout + nowTime, tcId, sequenceId);
+    }
+
+    @Override
+    public void start() {
+        run(null);
+    }
+
+    @Override
+    public void close() {
+        priorityQueue.close();
+        this.close();
+    }
+
+    @Override
+    public void run(Timeout timeout) {
+        synchronized (this){
+            while (!priorityQueue.isEmpty()){
+                long timeoutTime = priorityQueue.peekN1();
+                long nowTime = clock.millis() / BASE_OF_MILLIS_TO_SECOND;
+                if (timeoutTime < nowTime){
+                    transactionMetadataStoreService.endTransaction(new TxnID(priorityQueue.peekN2(),
```

Comment:
How to handle the transaction is committed?  we don't need to abort a transaction that already committed or aborted.